File: api-gateway-service/src/server/headers.js
const HOP_BY_HOP_HEADERS = new Set([
  "connection",
  "keep-alive",
  "proxy-connection",
  "transfer-encoding",
  "upgrade",
  "te",
  "trailer",
  "proxy-authorization",
  "proxy-authenticate",
  "host",
  "content-length",
  "expect",
]);

export function sanitizeRequestHeaders(headers) {
  const result = {};
  for (const [key, value] of Object.entries(headers)) {
    if (!value) continue;
    const lowerKey = key.toLowerCase();
    if (HOP_BY_HOP_HEADERS.has(lowerKey)) continue;
    result[key] = Array.isArray(value) ? value.join(", ") : value;
  }
  return result;
}

export function sanitizeResponseHeaders(headers) {
  const result = {};
  headers.forEach((value, key) => {
    if (HOP_BY_HOP_HEADERS.has(key.toLowerCase())) return;
    result[key] = value;
  });
  return result;
}

export function sanitizeHeadersForCache(headers) {
  const filtered = { ...headers };
  for (const key of Object.keys(filtered)) {
    const lower = key.toLowerCase();
    if (lower === "set-cookie" || lower === "set-cookie2" || lower === "content-length") {
      delete filtered[key];
    }
  }
  return filtered;
}

export function findHeaderKey(headers, target) {
  const lowerTarget = target.toLowerCase();
  return Object.keys(headers).find((key) => key.toLowerCase() === lowerTarget) ?? null;
}

function normalizeAddress(address) {
  if (!address) return null;
  if (address.startsWith("::ffff:")) {
    return address.slice(7);
  }
  if (address === "::1") {
    return "127.0.0.1";
  }
  return address;
}

export function appendForwardedFor(headers, remoteAddress) {
  if (!remoteAddress) {
    return;
  }
  const normalized = normalizeAddress(remoteAddress);
  if (!normalized) {
    return;
  }
  const existingKey = findHeaderKey(headers, "x-forwarded-for");
  if (existingKey) {
    const parts = headers[existingKey]
      .split(",")
      .map((item) => item.trim())
      .filter((item) => item.length > 0);
    if (!parts.includes(normalized)) {
      parts.push(normalized);
    }
    headers[existingKey] = parts.join(", ");
  } else {
    headers["X-Forwarded-For"] = normalized;
  }
}

export function resolveClientIp(req) {
  const headerCandidates = ["cf-connecting-ip", "cf-connection-ip"];
  for (const header of headerCandidates) {
    const raw = req.headers[header];
    const value = Array.isArray(raw) ? raw[0] : raw;
    if (typeof value === "string" && value.trim().length > 0) {
      const normalized = normalizeAddress(value.trim());
      if (normalized) {
        return { ip: normalized, source: header };
      }
    }
  }

  const rawForwarded = req.headers["x-forwarded-for"];
  const forwardedValue = Array.isArray(rawForwarded) ? rawForwarded[0] : rawForwarded;
  if (forwardedValue && forwardedValue.trim().length > 0) {
    const first = forwardedValue.split(",")[0].trim();
    const normalized = normalizeAddress(first);
    if (normalized) {
      return { ip: normalized, source: "x-forwarded-for" };
    }
  }

  const socketAddress = normalizeAddress(req.socket?.remoteAddress ?? null);
  if (socketAddress) {
    return { ip: socketAddress, source: "remote-address" };
  }

  return { ip: null, source: null };
}

File: api-gateway-service/src/headers.rs
use http::{HeaderMap, HeaderName, HeaderValue};
use std::net::SocketAddr;

const HOP_BY_HOP_HEADERS: [&str; 12] = [
    "connection",
    "keep-alive",
    "proxy-connection",
    "transfer-encoding",
    "upgrade",
    "te",
    "trailer",
    "proxy-authorization",
    "proxy-authenticate",
    "host",
    "content-length",
    "expect",
];

pub fn sanitize_request_headers(headers: &HeaderMap) -> HeaderMap {
    let mut result = HeaderMap::new();
    for (key, value) in headers.iter() {
        let lower = key.as_str().to_ascii_lowercase();
        if HOP_BY_HOP_HEADERS.contains(&lower.as_str()) {
            continue;
        }
        result.insert(key.clone(), value.clone());
    }
    result
}

pub fn sanitize_response_headers(headers: &HeaderMap) -> HeaderMap {
    let mut result = HeaderMap::new();
    for (key, value) in headers.iter() {
        let lower = key.as_str().to_ascii_lowercase();
        if HOP_BY_HOP_HEADERS.contains(&lower.as_str()) {
            continue;
        }
        result.insert(key.clone(), value.clone());
    }
    result
}

pub fn sanitize_headers_for_cache(headers: &HeaderMap) -> HeaderMap {
    let mut filtered = HeaderMap::new();
    for (key, value) in headers.iter() {
        let lower = key.as_str().to_ascii_lowercase();
        if lower == "set-cookie" || lower == "set-cookie2" || lower == "content-length" {
            continue;
        }
        filtered.insert(key.clone(), value.clone());
    }
    filtered
}

pub fn find_header_key(headers: &HeaderMap, target: &str) -> Option<HeaderName> {
    let target_lower = target.to_ascii_lowercase();
    headers
        .keys()
        .find(|name| name.as_str().eq_ignore_ascii_case(&target_lower))
        .cloned()
}

fn normalize_address(value: &str) -> Option<String> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return None;
    }
    if let Some(stripped) = trimmed.strip_prefix("::ffff:") {
        return Some(stripped.to_string());
    }
    if trimmed == "::1" {
        return Some("127.0.0.1".into());
    }
    Some(trimmed.to_string())
}

pub fn append_forwarded_for(headers: &mut HeaderMap, remote: Option<&SocketAddr>) {
    let ip = remote.map(|addr| addr.ip().to_string());
    let Some(ip) = ip else {
        return;
    };

    if let Some(existing_key) = find_header_key(headers, "x-forwarded-for")
        && let Some(existing) = headers.get_mut(&existing_key)
    {
        let mut parts: Vec<String> = existing
            .to_str()
            .unwrap_or_default()
            .split(',')
            .map(|item| item.trim().to_string())
            .filter(|item| !item.is_empty())
            .collect();
        if !parts.iter().any(|item| item == &ip) {
            parts.push(ip.clone());
        }
        if let Ok(value) = HeaderValue::from_str(&parts.join(", ")) {
            *existing = value;
        }
        return;
    }

    if let Ok(value) = HeaderValue::from_str(&ip) {
        headers.insert(HeaderName::from_static("x-forwarded-for"), value);
    }
}

pub struct ClientIp {
    pub ip: Option<String>,
    #[allow(dead_code)]
    pub source: Option<&'static str>,
}

pub fn resolve_client_ip(
    headers: &HeaderMap,
    remote: Option<&SocketAddr>,
    trust_proxy: bool,
) -> ClientIp {
    if trust_proxy {
        for header in ["cf-connecting-ip", "cf-connection-ip"] {
            if let Some(value) = headers.get(header).and_then(|value| value.to_str().ok())
                && let Some(ip) = normalize_address(value)
            {
                return ClientIp {
                    ip: Some(ip),
                    source: Some(header),
                };
            }
        }

        if let Some(value) = headers
            .get("x-forwarded-for")
            .and_then(|value| value.to_str().ok())
            && let Some(first) = value.split(',').next().map(|item| item.trim())
            && let Some(ip) = normalize_address(first)
        {
            return ClientIp {
                ip: Some(ip),
                source: Some("x-forwarded-for"),
            };
        }
    }

    let socket_addr = remote.map(|addr| addr.ip().to_string());
    let source = socket_addr.as_ref().map(|_| "remote-address");
    ClientIp {
        ip: socket_addr,
        source,
    }
}
